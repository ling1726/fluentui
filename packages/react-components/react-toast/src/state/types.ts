import { EVENTS } from './constants';

export type ToastId = string;

export type ToastPosition = 'top-right' | 'top-center' | 'top-left' | 'bottom-right' | 'bottom-center' | 'bottom-left';

export interface ToastOptions {
  toastId?: ToastId;
  position?: ToastPosition;
  content?: unknown;
  timeout?: number;
  pauseOnWindowBlur?: boolean;
  pauseOnHover?: boolean;
}

export interface ToasterOptions extends DefaultToastOptions {}

export interface DefaultToastOptions
  extends Pick<ToastOptions, 'position' | 'timeout' | 'pauseOnWindowBlur' | 'pauseOnHover'> {}

export interface ValidatedToastOptions extends Required<DefaultToastOptions> {}

export interface Toast extends Required<ToastOptions> {
  close: () => void;
  remove: () => void;
}

export interface DismissToastEventDetail {
  toastId: ToastId | undefined;
}

export interface ToastEventMap {
  [EVENTS.show]: CustomEvent<ToastOptions>;
  [EVENTS.dismiss]: CustomEvent<DismissToastEventDetail>;
}

export type ToastEventListenerGeneric<K extends keyof ToastEventMap> = (e: ToastEventMap[K]) => void;
export type ToastEventListener = <K extends keyof ToastEventMap>(e: ToastEventMap[K]) => void;
