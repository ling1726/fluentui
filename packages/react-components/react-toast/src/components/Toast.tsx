/**
 * ⚠️ This is temporary and WILL be removed
 */

import * as React from 'react';
import { Transition } from 'react-transition-group';
import { useIsomorphicLayoutEffect, useMergedRefs } from '@fluentui/react-utilities';
import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { useToast, ToastProps } from '../state';
import { Timer } from './Timer';

const useStyles = makeStyles({
  toast: {
    ...shorthands.border('2px', 'dashed', 'red'),
    ...shorthands.padding('4px'),
    display: 'flex',
    minHeight: '40px',
    maxHeight: '40px',
    minWidth: '200px',
    maxWidth: '200px',
    alignItems: 'center',
    backgroundColor: 'white',
  },

  slide: {
    animationDuration: '200ms, 400ms',
    animationDelay: '0ms, 200ms',
    animationName: [
      {
        from: {
          height: '0',
          minHeight: '0',
          maxHeight: '0',
          opacity: 0,
        },
        to: {
          opacity: 0,
        },
      },
      {
        from: {
          opacity: 0,
        },
        to: {
          opacity: 1,
        },
      },
    ],
  },

  fadeOut: {
    animationDuration: '400ms, 200ms',
    animationDelay: '0ms, 400ms',
    animationName: [
      {
        from: {
          opacity: 1,
        },
        to: {
          opacity: 0,
        },
      },
      {
        from: {
          opacity: 0,
        },
        to: {
          opacity: 0,
          height: 0,
          maxHeight: 0,
          minHeight: 0,
        },
      },
    ],
  },
});

export const Toast: React.FC<ToastProps> = props => {
  const styles = useStyles();
  const { isIn, children, closeToast, deleteToast, autoClose } = props;
  const { eventHandlers, toastRef, playToast, isRunning } = useToast(props);
  const nodeRef = React.useRef<HTMLDivElement>(null);

  // start the toast once it's fully in
  useIsomorphicLayoutEffect(() => {
    if (toastRef.current) {
      const toast = toastRef.current;
      toast.addEventListener('animationend', playToast, {
        once: true,
      });

      return () => {
        toast.removeEventListener('animationend', playToast);
      };
    }
  }, [playToast, toastRef]);

  return (
    <Transition in={isIn} unmountOnExit mountOnEnter timeout={500} onExited={deleteToast} nodeRef={nodeRef}>
      <div
        ref={useMergedRefs(nodeRef, toastRef)}
        className={mergeClasses(styles.toast, isIn && styles.slide, !isIn && styles.fadeOut)}
        {...eventHandlers}
      >
        {children as React.ReactNode}
        <Timer onTimeout={closeToast} timeout={!autoClose ? -1 : autoClose} isRunning={isRunning} />
      </div>
    </Transition>
  );
};
