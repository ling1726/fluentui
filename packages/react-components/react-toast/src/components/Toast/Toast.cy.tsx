import * as React from 'react';
import { mount as mountBase } from '@cypress/react';

import { FluentProvider } from '@fluentui/react-provider';
import { teamsLightTheme } from '@fluentui/react-theme';
import { Toaster, ToastTitle, Toast, useToastController, toastClassNames } from '../..';

const mount = (element: JSX.Element) => {
  mountBase(<FluentProvider theme={teamsLightTheme}>{element}</FluentProvider>);
};

describe('Toast', () => {
  it('should dispatch toast', () => {
    const Example = () => {
      const { dispatchToast } = useToastController();
      const onClick = () => {
        dispatchToast(
          <Toast>
            <ToastTitle>This is a toast</ToastTitle>
          </Toast>,
        );
      };

      return (
        <>
          <button onClick={onClick}>Make toast</button>
          <Toaster />
        </>
      );
    };

    mount(<Example />);
    cy.get('button').click().get(`.${toastClassNames.root}`).should('exist');
  });

  it('should dismiss toast', () => {
    const Example = () => {
      const toastId = 'foo';
      const { dispatchToast, dismissToast } = useToastController();
      const makeToast = () => {
        dispatchToast(
          <Toast>
            <ToastTitle>This is a toast</ToastTitle>
          </Toast>,
          { toastId, timeout: -1 },
        );
      };

      const removeToast = () => {
        dismissToast(toastId);
      };

      return (
        <>
          <button id="make" onClick={makeToast}>
            Make toast
          </button>
          <button id="dismiss" onClick={removeToast}>
            Dismiss toast
          </button>
          <Toaster />
        </>
      );
    };

    mount(<Example />);
    cy.get('#make')
      .click()
      .get(`.${toastClassNames.root}`)
      .should('exist')
      .get('#dismiss')
      .click()
      .get(`.${toastClassNames.root}`)
      .should('not.exist');
  });

  it('should dismiss all toasts', () => {
    const Example = () => {
      const { dispatchToast, dismissAllToasts } = useToastController();
      const makeToast = () => {
        for (let i = 0; i < 5; i++) {
          dispatchToast(
            <Toast>
              <ToastTitle>This is a toast</ToastTitle>
            </Toast>,
            { timeout: -1 },
          );
        }
      };

      const removeToast = () => {
        dismissAllToasts();
      };

      return (
        <>
          <button id="make" onClick={makeToast}>
            Make toast
          </button>
          <button id="dismiss" onClick={removeToast}>
            Dismiss toast
          </button>
          <Toaster />
        </>
      );
    };

    mount(<Example />);
    cy.get('#make')
      .click()
      .get(`.${toastClassNames.root}`)
      .should('have.length', 5)
      .get('#dismiss')
      .click()
      .get(`.${toastClassNames.root}`)
      .should('not.exist');
  });

  it('should be update toast', () => {
    const Example = () => {
      const toastId = 'foo';
      const { dispatchToast, updateToast } = useToastController();
      const makeToast = () => {
        for (let i = 0; i < 5; i++) {
          dispatchToast(
            <Toast>
              <ToastTitle>This is a toast</ToastTitle>
            </Toast>,
            { timeout: -1, toastId },
          );
        }
      };

      const update = () => {
        updateToast({
          content: (
            <Toast>
              <ToastTitle>Foo</ToastTitle>
            </Toast>
          ),
          toastId,
        });
      };

      return (
        <>
          <button id="make" onClick={makeToast}>
            Make toast
          </button>
          <button id="update" onClick={update}>
            Update toast
          </button>
          <Toaster />
        </>
      );
    };

    mount(<Example />);
    cy.get('#make')
      .click()
      .get(`.${toastClassNames.root}`)
      .should('exist')
      .get('#update')
      .click()
      .get('body')
      .contains('Foo');
  });

  it('should be pause on hover', () => {
    const Example = () => {
      const { dispatchToast } = useToastController();
      const makeToast = () => {
        dispatchToast(
          <Toast>
            <ToastTitle>This is a toast</ToastTitle>
          </Toast>,
          { timeout: 500, pauseOnHover: true },
        );
      };

      return (
        <>
          <button id="make" onClick={makeToast}>
            Make toast
          </button>
          <Toaster />
        </>
      );
    };

    mount(<Example />);
    cy.get('#make')
      .click()
      .get(`.${toastClassNames.root}`)
      .trigger('mouseenter')
      .wait(700)
      .get(`.${toastClassNames.root}`)
      .should('exist');
  });

  it('should be pause all toasts on focus', () => {
    const Example = () => {
      const { dispatchToast } = useToastController();
      const makeToast = () => {
        for (let i = 0; i < 4; i++) {
          dispatchToast(
            <Toast>
              <ToastTitle>This is a toast</ToastTitle>
            </Toast>,
            { timeout: 500, pauseOnHover: true },
          );
        }

        dispatchToast(
          <Toast>
            <ToastTitle action={<button id="action">action</button>}>This is a toast</ToastTitle>
          </Toast>,
          { timeout: 500, pauseOnHover: true },
        );
      };

      return (
        <>
          <button id="make" onClick={makeToast}>
            Make toast
          </button>
          <Toaster />
        </>
      );
    };

    mount(<Example />);
    cy.get('#make').click().get(`#action`).focus().wait(700).get(`.${toastClassNames.root}`).should('have.length', 5);
  });
});
