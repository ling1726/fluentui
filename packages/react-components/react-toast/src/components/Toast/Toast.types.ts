import * as React from 'react';
import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
import { Announce } from '../AriaLive/AriaLive.types';
import { Toast as VanillaToast } from '../../state';
import { ToastContextValue } from '../../contexts/toastContext';

export type ToastContextValues = {
  toast: ToastContextValue;
};

export type ToastSlots = {
  root: Slot<'div'>;
};

/**
 * Toast Props
 */
export type ToastProps = ComponentProps<ToastSlots> &
  VanillaToast & {
    visible: boolean;
    announce: Announce;
  };

/**
 * State used in rendering Toast
 */
export type ToastState = ComponentState<ToastSlots> &
  Pick<ToastProps, 'remove' | 'close' | 'updateId' | 'visible'> & {
    transitionTimeout: number;
    timerTimeout: number;
    running: boolean;
    onTransitionEntering: () => void;
    nodeRef: React.Ref<HTMLDivElement>;
  };
