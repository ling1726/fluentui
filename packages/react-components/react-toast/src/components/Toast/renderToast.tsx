/** @jsxRuntime classic */
/** @jsx createElement */

import { createElement } from '@fluentui/react-jsx-runtime';
import { getSlotsNext } from '@fluentui/react-utilities';
import { BackgroundAppearanceProvider } from '@fluentui/react-shared-contexts';
import type { ToastState, ToastSlots, ToastContextValues } from './Toast.types';

/**
 * Render the final JSX of Toast
 */
export const renderToast_unstable = (state: ToastState, contexts: ToastContextValues) => {
  const { slots, slotProps } = getSlotsNext<ToastSlots>(state);

  return (
    <BackgroundAppearanceProvider value={contexts.backgroundAppearance}>
      <slots.root {...slotProps.root} />
    </BackgroundAppearanceProvider>
  );
};
