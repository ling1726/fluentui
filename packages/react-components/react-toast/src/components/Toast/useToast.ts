import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import type { ToastProps, ToastState } from './Toast.types';
import { useToastContext } from '../../contexts/toastContext';

/**
 * Create the state required to render Toast.
 *
 * The returned state can be modified with hooks such as useToastStyles_unstable,
 * before being passed to renderToast_unstable.
 *
 * @param props - props from this instance of Toast
 * @param ref - reference to root HTMLElement of Toast
 */
export const useToast_unstable = (props: ToastProps, ref: React.Ref<HTMLElement>): ToastState => {
  const { titleId } = useToastContext();
  return {
    components: {
      root: 'div',
    },
    root: getNativeElementProps('div', {
      ref,
      'aria-labelledby': titleId,
      ...props,
    }),
  };
};
