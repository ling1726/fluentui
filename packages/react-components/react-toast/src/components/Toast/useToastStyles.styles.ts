import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import { tokens } from '@fluentui/react-theme';
import type { ToastSlots, ToastState } from './Toast.types';
import type { SlotClassNames } from '@fluentui/react-utilities';

export const toastClassNames: SlotClassNames<ToastSlots> = {
  root: 'fui-Toast',
  timer: 'fui-Timer',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    boxSizing: 'border-box',
    marginTop: '16px',
    minHeight: '44px',
    pointerEvents: 'all',
    ...shorthands.borderRadius(tokens.borderRadiusMedium),
    '--fui-toast-height': '44px',
  },

  enter: {
    animationDuration: '200ms, 400ms',
    animationDelay: '0ms, 200ms',
    animationName: [
      {
        from: {
          maxHeight: 0,
          opacity: 0,
          marginTop: 0,
        },
        to: {
          marginTop: '16px',
          opacity: 0,
          maxHeight: 'var(--fui-toast-height)',
        },
      },
      {
        from: {
          opacity: 0,
        },
        to: {
          opacity: 1,
        },
      },
    ],
  },

  exit: {
    animationDuration: '400ms, 200ms',
    animationDelay: '0ms, 400ms',
    animationName: [
      {
        from: {
          opacity: 1,
        },
        to: {
          opacity: 0,
        },
      },
      {
        from: {
          opacity: 0,
        },
        to: {
          opacity: 0,
          marginTop: 0,
          maxHeight: 0,
        },
      },
    ],
  },
});

/**
 * Apply styling to the Toast slots based on the state
 */
export const useToastStyles_unstable = (state: ToastState): ToastState => {
  const styles = useStyles();
  state.root.className = mergeClasses(
    toastClassNames.root,
    state.visible ? styles.enter : styles.exit,
    styles.root,
    state.root.className,
  );

  return state;
};
