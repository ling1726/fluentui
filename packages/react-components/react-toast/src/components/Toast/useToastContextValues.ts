import * as React from 'react';
import { ToastContextValues, ToastState } from './Toast.types';

export function useToastContextValues_unstable(state: ToastState): ToastContextValues {
  const { close } = state;

  const toastContext = React.useMemo(
    () => ({
      close,
    }),
    [close],
  );

  return {
    toast: toastContext,
  };
}
