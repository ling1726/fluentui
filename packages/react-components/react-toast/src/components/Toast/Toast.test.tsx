import * as React from 'react';
import { render, act, fireEvent } from '@testing-library/react';
import { Toast } from './Toast';
import { isConformant } from '../../testing/isConformant';
import { ToastProps } from './Toast.types';
import { toastClassNames } from './useToastStyles.styles';

const defaultToastProps: ToastProps = {
  announce: () => null,
  close: () => null,
  content: '',
  data: {},
  dispatchedAt: 0,
  pauseOnHover: false,
  pauseOnWindowBlur: false,
  politeness: 'polite',
  position: 'bottom-left',
  priority: 0,
  remove: () => null,
  timeout: -1,
  toasterId: 'toaster',
  toastId: 'toast',
  updateId: 0,
  visible: true,
};

const runningTimerSelector = '[data-timer-status="running"]';
const pausedTimerSelector = '[data-timer-status="paused"]';

describe('Toast', () => {
  beforeEach(() => {
    jest.useRealTimers();
  });

  isConformant({
    Component: Toast,
    displayName: 'Toast',
    requiredProps: defaultToastProps,
  });

  it('renders a default state', () => {
    const result = render(<Toast {...defaultToastProps}>Default Toast</Toast>);
    expect(result.container).toMatchSnapshot();
  });

  it('should announce on mount', () => {
    const announce = jest.fn();
    const toastProps = { ...defaultToastProps, announce };
    render(<Toast {...toastProps}>Toast</Toast>);

    expect(announce).toHaveBeenCalledTimes(1);
    expect(announce).toHaveBeenCalledWith('Toast', { politeness: 'polite' });
  });

  it('should announce on updateId change', () => {
    const announce = jest.fn();
    const toastProps = { ...defaultToastProps, announce };
    const { rerender } = render(<Toast {...toastProps}>Toast</Toast>);

    toastProps.updateId++;

    rerender(<Toast {...toastProps}>Toast</Toast>);

    expect(announce).toHaveBeenCalledTimes(2);
    expect(announce).toHaveBeenNthCalledWith(2, 'Toast', { politeness: 'polite' });
  });

  it('should announce with configured politeness', () => {
    const announce = jest.fn();
    const toastProps: ToastProps = { ...defaultToastProps, announce, politeness: 'assertive' };
    render(<Toast {...toastProps}>Toast</Toast>);

    expect(announce).toHaveBeenCalledTimes(1);
    expect(announce).toHaveBeenCalledWith('Toast', { politeness: 'assertive' });
  });

  it('should respect user root', () => {
    const className = 'foo';
    const toastProps: ToastProps = { ...defaultToastProps, data: { root: { className } } };
    const { container } = render(<Toast {...toastProps}>Toast</Toast>);

    expect(container.querySelector(`.${className}`)).not.toBe(null);
  });

  it('should remove toast after visible is false', () => {
    jest.useFakeTimers();
    const remove = jest.fn();
    const toastProps: ToastProps = { ...defaultToastProps, remove };
    const { rerender } = render(<Toast {...toastProps}>Toast</Toast>);
    toastProps.visible = false;
    rerender(<Toast {...toastProps}>Toast</Toast>);

    jest.advanceTimersByTime(500);

    expect(remove).toHaveBeenCalledTimes(1);
  });

  it('should start timer after toast on animationend', () => {
    const toastProps: ToastProps = { ...defaultToastProps, timeout: 1 };
    const { container } = render(<Toast {...toastProps}>Toast</Toast>);

    const toastElement = container.querySelector(`.${toastClassNames.root}`);
    expect(toastElement).not.toBeNull();
    act(() => {
      if (toastElement) {
        fireEvent.animationEnd(toastElement);
      }
    });

    expect(container.querySelector('[data-timer-status="running"]')).not.toBeNull();
  });

  it('should close toast ontimeout', () => {
    const close = jest.fn();
    const toastProps: ToastProps = { ...defaultToastProps, timeout: 1, close };
    const { container } = render(<Toast {...toastProps}>Toast</Toast>);

    const toastElement = container.querySelector(`.${toastClassNames.root}`);
    expect(toastElement).not.toBeNull();
    act(() => {
      toastElement?.dispatchEvent(new Event('animationend'));
    });

    const timer = container.querySelector(runningTimerSelector);
    expect(timer).not.toBeNull();

    act(() => {
      if (timer) {
        fireEvent.animationEnd(timer);
      }
    });

    expect(close).toHaveBeenCalledTimes(1);
  });

  it('should pause on hover', () => {
    const toastProps: ToastProps = { ...defaultToastProps, timeout: 1, pauseOnHover: true };
    const { container } = render(<Toast {...toastProps}>Toast</Toast>);

    const toastElement = container.querySelector(`.${toastClassNames.root}`);
    expect(toastElement).not.toBeNull();
    act(() => {
      if (toastElement) {
        fireEvent.animationEnd(toastElement);
      }
    });

    expect(container.querySelector(runningTimerSelector)).not.toBeNull();

    act(() => {
      if (toastElement) {
        fireEvent.mouseEnter(toastElement);
      }
    });

    expect(container.querySelector(pausedTimerSelector)).not.toBeNull();

    act(() => {
      if (toastElement) {
        fireEvent.mouseLeave(toastElement);
      }
    });

    expect(container.querySelector(runningTimerSelector)).not.toBeNull();
  });

  it('should pause on window blur', () => {
    const toastProps: ToastProps = { ...defaultToastProps, timeout: 1, pauseOnWindowBlur: true };
    const { container } = render(<Toast {...toastProps}>Toast</Toast>);

    const toastElement = container.querySelector(`.${toastClassNames.root}`);
    expect(toastElement).not.toBeNull();
    act(() => {
      if (toastElement) {
        fireEvent.animationEnd(toastElement);
      }
    });

    expect(container.querySelector(runningTimerSelector)).not.toBeNull();

    act(() => {
      fireEvent.blur(window);
    });

    expect(container.querySelector(pausedTimerSelector)).not.toBeNull();

    act(() => {
      fireEvent.focus(window);
    });

    expect(container.querySelector(runningTimerSelector)).not.toBeNull();
  });
});
