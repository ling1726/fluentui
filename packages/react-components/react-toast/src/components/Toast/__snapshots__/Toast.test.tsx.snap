// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast renders a default state 1`] = `
<div>
  <div
    aria-labelledby=""
    class="fui-Toast"
  >
    Default Toast
  </div>
</div>
`;
